src/Featurizers/PCAFeaturizer.h

```diff
@@ -0,0 +1,267 @@
+
+            // Data
+            MatrixT                        eigenvalues(Traits<MatrixT>::deserialize(ar));
+            MatrixT                        eigenvectors(Traits<MatrixT>::deserialize(ar));
+
+            return PCATransformer<MatrixT>(std::move(eigenvalues), std::move(eigenvectors));
+        }()
+    ) {
+}
+
+template <typename MatrixT>
+bool PCATransformer<MatrixT>::operator==(PCATransformer const &other) const {
+    if ((this->_eigenvalues - other._eigenvalues).norm() > 0.000001f || (this->_eigenvectors - other._eigenvectors).norm() > 0.000001f)
+        return false;
+
+    return true;
+}
+
+template <typename MatrixT>
+void PCATransformer<MatrixT>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<decltype(_eigenvalues)>::serialize(ar, _eigenvalues);
+    Traits<decltype(_eigenvectors)>::serialize(ar, _eigenvectors);
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  PCAEstimator
+// |
+// ----------------------------------------------------------------------
+template <typename MatrixT, size_t MaxNumTrainingItemsV>
+PCAEstimator<MatrixT, MaxNumTrainingItemsV>::PCAEstimator(AnnotationMapsPtr pAllColumnAnnotations, size_t colIndex) :
+    BaseType("SVDEstimatorImpl", std::move(pAllColumnAnnotations)),
+    _colIndex(
+        std::move(
+            [this, &colIndex](void) -> size_t & {
+                if(colIndex >= this->get_column_annotations().size())
+                    throw std::invalid_argument("colIndex");
+
+                return colIndex;
+            }()
+        )
+    ) {
+        _hasCalledFit = false;
+}
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+template <typename MatrixT,size_t MaxNumTrainingItemsV>
+bool PCAEstimator<MatrixT, MaxNumTrainingItemsV>::begin_training_impl(void) /*override*/ {
+    return true;
+}
+
+template <typename MatrixT, size_t MaxNumTrainingItemsV>
+void PCAEstimator<MatrixT, MaxNumTrainingItemsV>::complete_training_impl(void) /*override*/ {
+
+    Eigen::SelfAdjointEigenSolver<MatrixT> eig(
+        [&]() -> MatrixT {
+            //compute for centered training data
+            MatrixT centered = _matrix.rowwise() - _matrix.colwise().mean();
+            //compute for covariance matrix
+            MatrixT cov = centered.adjoint() * centered;
+            //using covariance matrix to compute PCA
+            return cov;
+        }()
+    );
+
+    // free _matrix
+    _matrix.resize(0, 0);
+
+    //PCA uses sigma but sklearn PCA uses sqrt(sigma), this implementation follows original principle.
+    this->_eigenvalues = eig.eigenvalues();
+    this->_eigenvectors = eig.eigenvectors();
```

Comment:
Is it better to invoke this functionality here or when the transformer is created? If you move it to create_transformer_impl, you can remove _eigenvalues and _eigenvectors as members of the class.